### ISSUE TITLE
sphinx-build -W treats non-sphinx warnings as errors

### ISSUE THREAD

**Issue by rouge8** [sphinx-doc/sphinx#3926]
### Problem

`sphinx-build -W` treats non-sphinx warnings as errors. I've included a minimal reproduction. It seems like this is due to https://github.com/sphinx-doc/sphinx/pull/3267 because it is new to 1.6+

#### Procedure to reproduce the problem

```
sphinx-build -b html . _build -W
```

#### Error logs / results

```
Running Sphinx v1.6.3
loading pickled environment... not yet created
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 1 source files that are out of date
updating environment: 1 added, 0 changed, 0 removed
reading sources... [100%] index

Warning, treated as error:
sphinx should not warn about this
```

#### Expected results

Build should succeed

### Reproducible project / your project

Minimal project: https://gist.github.com/rouge8/c10f6178c8f0dc235ce9440af0519a99

### Environment info
- OS: OS X 10.11.6
- Python version: Python 2.7.13
- Sphinx version: Sphinx (sphinx-build) 1.6.3

**Comment by stephenfin** [comment]
I don't think this is really an issue. You're using the [`sphinx.ext.autodoc`](http://www.sphinx-doc.org/en/stable/ext/autodoc.html) extension. According to the docs for same:

> autodoc imports the modules to be documented. If any modules have side effects on import, these will be executed by autodoc when sphinx-build is run.
> If you document scripts (as opposed to library modules), make sure their main routine is protected by a if __name__ == '__main__' condition.

If you don't want warnings, then you should either exclude that module (by way of arguments to the [`sphinx-apidoc`](http://www.sphinx-doc.org/en/stable/man/sphinx-apidoc.html) command) or move the warning into a function.

Also note that "fixing" this (in quotes because I don't think it's an issue), will break every plugin that doesn't inherit from some type of Sphinx warning, which would be all of them (because we don't have such a warning class at the moment).

**Comment by rouge8** [comment]
How did plugins behave before 1.6? Our actual error is using newrelic's Python agent, where a decorator attempts to activate the agent and warns there.

This works fine in 1.5.6 (and I can patch the offending function on 1.6.x) so feel free to close if this is intended behavior.

**Comment by stephenfin** [comment]
> Our actual error is using newrelic's Python agent, where a decorator attempts to activate the agent and warns there.
> 
> This works fine in 1.5.6 (and I can patch the offending function on 1.6.x) so feel free to close if this is intended behavior.

Ah, I missed the reference to #3267. Hmm, that's a trickier one than I first though.

Given that `sphinx.ext.autodoc` is involved here, I'd _still_ be inclined to think this is not a bug. However, I am not as certain of this as I was before. Guess we need more input from others.

**Comment by stephenfin** [comment]
> How did plugins behave before 1.6?

Based on the few plugins I've looked at, mostly using a combo of `docutils.parsers.rst.Directive.error` and `warnings.warning`. These presumably wouldn't be affected, as the change to logging hasn't broken them per se.

However, I guess the only way to fix this would be to start filtering on the loggers that we raise an error on. To do this, we'd probably have to define some pattern that loggers has to follow, or add some form of "register logger" command Sphinx. In either case, we probably would break plugins written *since* #3267 as these would hopefully be using `logger.warning`.

**Comment by tk0miya** [comment]
Thank for reporting, and sorry for inconvenience.
Surely, this is undesirable behavior.

It would be better to disable `-W` option during importing targets scripts of autodoc.

**Comment by tk0miya** [comment]
Now I send #3931 to resolve this. Please try it out if possible.

**Comment by rouge8** [comment]
@tk0miya #3931 works great, thanks! 👍

**Comment by tk0miya** [comment]
Thank you for confirmation. Merged.
It will be released as 1.6.4 soon :-)
